### ISSUE TITLE
RFC: add get_request_handler(const char *uri, ....) to public api

### ISSUE THREAD

**Issue by PhilLab** [civetweb/civetweb#731]
The current ``get_request_handler`` is internal. I propose to provide a similar method which takes not a connection but an uri and returns the handler for it. If we are on it, it should also be available in the C++ bindings.

This allows a user to access registered handlers for lookup/modification without having an actual connection open. My actual use case is that I have custom handlers registered (``CivetServer::addWebSocketHandler()``) and would want to list them all using civetweb directly.

**Comment by bel2125** [comment]
Currently the connection is used as an input, because `get_request_handler` needs the context (domain context for URIs and "physical" context for locks and memory) - this can be accessed from the "connection" object.
I just checked, it is also possible to find it only from the server context returned from mg_start, but you will need the host name part from the URI as well (in contrast to all other APIs using URIs).

Getting a list of all registered URIs to get all handlers will require a different interface.

Did you already think about a C interface that could suit your needs?

**Comment by PhilLab** [comment]
A useful C interface (at least for me) would be something like
```
get_request_handler(struct mg_context *ctx,
                    const char *uri, int handler_type, struct mg_handler_info **handler_info)
```
I am not that into C interfaces, so I am unsure if the return parameter ``**handler_info`` is sufficient or if we need all handler pointers:
```
get_request_handler(struct mg_context *ctx,
                    const char *uri, int handler_type,
                    mg_request_handler *handler,
                    struct mg_websocket_subprotocols **subprotocols,
                    mg_websocket_connect_handler *connect_handler,
                    mg_websocket_ready_handler *ready_handler,
                    mg_websocket_data_handler *data_handler,
                    mg_websocket_close_handler *close_handler,
                    mg_authorization_handler *auth_handler,
                    void **cbdata,
                    struct mg_handler_info **handler_info)
```

In the C++ interface, we could use the context from mg_start as it is done on other places in CivetServer.cpp, ``me->context``

**Comment by bel2125** [comment]
Don't worry about the details of the C interfaces - the reason I was asking for a C interface proposal was to get a short form of what inputs and outputs you expect/need. So the first form is sufficient for you (yes it can also be done in C without adding unused pointers). Actually `struct mg_handler_info` contains all these function pointers anyway:

https://github.com/civetweb/civetweb/blob/b21ca4a0a6b54c1ee1223be1f6a823640f1b2c50/src/civetweb.c#L2536-L2568

So the `get_request_handler` call became redundant during development when merging different contributions - I can restructure this some day (not today :wink:).

One issue is, `struct mg_handler_info` is an internal structure, and probably extended by some elements (eg. this issue https://github.com/civetweb/civetweb/issues/667#issuecomment-460805699), maybe others get dropped/renamed when something is refactored.

Would it be sufficient to return for every URL (including hostname and protocol), if there is a handler or not, and return `void *cbdata;` of this handler?

**Comment by PhilLab** [comment]
Yes, returning cbdata or null would also be sufficient. The only thing I am thinking about if the CPP interface could do some casting to be more "c++-style" and to match the already existing calls

(Pseudocode)
```
CivetHandler*
CivetServer::getConnectionHandler(const std::string &uri)
{
    get_request_handler(uri, handler_type_for_websockets, cbdata);
    return (CivetHandler*)cbdata
}

CivetWebSocketHandler*
CivetServer::getWebSocketConnectionHandler(const std::string &uri)
{
    get_request_handler(uri, handler_type_for_websockets, cbdata);
    return (CivetWebSocketHandler*)cbdata
}
```

**Comment by bel2125** [comment]
The `void *cbdata` would be the user supplied argument in the call to `mg_set_request_handler` or `mg_set_websocket_handler` or `mg_set_auth_handler`:

https://github.com/civetweb/civetweb/blob/38fe585a5c4f6584c59e31eb9c8ac78efbd09fc0/include/civetweb.h#L479-L482

https://github.com/civetweb/civetweb/blob/38fe585a5c4f6584c59e31eb9c8ac78efbd09fc0/include/civetweb.h#L535-L541

https://github.com/civetweb/civetweb/blob/38fe585a5c4f6584c59e31eb9c8ac78efbd09fc0/include/civetweb.h#L577-L580

So, this parameter is of type `void *` for the C interface - for the C++ wrapper this is a `CivetHandler *`, ... as you wrote.

https://github.com/civetweb/civetweb/blob/38fe585a5c4f6584c59e31eb9c8ac78efbd09fc0/src/CivetServer.cpp#L351-L354

**Comment by bel2125** [comment]
Is there still anything left to do, or can we close this issue?
